File: bac-simulator-web/lib/bac-calculator.ts
/**
 * BAC Calculation Engine - Widmark Equation with Food Absorption Model
 * TypeScript port of scientifically accurate blood alcohol content simulator
 *
 * BAC Formula: BAC = [(A × 5.14) / (W × r)] - (0.015 × H)
 * Where: A = alcohol consumed (oz), W = weight (lbs), r = Widmark ratio, H = hours
 */

import {
  UserProfile,
  Drink,
  Food,
  DrinkType,
  FoodType,
  BACResult,
  ImpairmentLevel,
  TimelinePoint,
  BACPeak,
  STANDARD_DRINKS,
  FOOD_GASTRIC_TIMES,
  FOOD_ABSORPTION_IMPACT,
  WIDMARK_RATIOS,
  IMPAIRMENT_LEVELS,
} from './types';

// Elimination rate: % BAC per hour (15 mg/100mL per hour ≈ 0.015%)
const ELIMINATION_RATE = 0.015;

export class BACCalculator {
  private drinks: Drink[] = [];
  private foods: Food[] = [];
  private profile: UserProfile;
  private startTime: Date;

  constructor(profile?: Partial<UserProfile>) {
    this.profile = {
      sex: 'male',
      weightLbs: 180,
      heightInches: 70,
      age: 30,
      chronicDrinker: false,
      ...profile,
    };
    this.startTime = new Date();
  }

  /**
   * Set user profile for BAC calculations
   */
  setProfile(profile: Partial<UserProfile>): void {
    this.profile = { ...this.profile, ...profile };
  }

  /**
   * Get current profile
   */
  getProfile(): UserProfile {
    return { ...this.profile };
  }

  /**
   * Add food consumed to timeline
   */
  addFood(time: Date, type: FoodType): string {
    const id = `food_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
    this.foods.push({ id, time, type });
    this.foods.sort((a, b) => a.time.getTime() - b.time.getTime());
    return id;
  }

  /**
   * Add drink(s) consumed to timeline
   */
  addDrink(
    time: Date,
    type: DrinkType,
    sizeOz?: number,
    alcoholPercent?: number,
    quantity: number = 1
  ): string[] {
    const stdDrink = STANDARD_DRINKS[type];
    const finalSizeOz = sizeOz ?? stdDrink.oz;
    const finalAlcoholPercent = alcoholPercent ?? stdDrink.alcoholPercent;
    const ids: string[] = [];

    for (let i = 0; i < quantity; i++) {
      // Spread multiple drinks 30 minutes apart
      const drinkTime = new Date(time.getTime() + i * 30 * 60 * 1000);
      const id = `drink_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;

      this.drinks.push({
        id,
        time: drinkTime,
        type,
        sizeOz: finalSizeOz,
        alcoholPercent: finalAlcoholPercent,
      });
      ids.push(id);
    }

    this.drinks.sort((a, b) => a.time.getTime() - b.time.getTime());

    // Update start time if this is earlier
    if (this.drinks.length > 0) {
      const earliestDrink = this.drinks[0].time;
      if (earliestDrink < this.startTime) {
        this.startTime = earliestDrink;
      }
    }

    return ids;
  }

  /**
   * Remove a drink by ID
   */
  removeDrink(id: string): boolean {
    const index = this.drinks.findIndex(d => d.id === id);
    if (index !== -1) {
      this.drinks.splice(index, 1);
      return true;
    }
    return false;
  }

  /**
   * Remove food by ID
   */
  removeFood(id: string): boolean {
    const index = this.foods.findIndex(f => f.id === id);
    if (index !== -1) {
      this.foods.splice(index, 1);
      return true;
    }
    return false;
  }

  /**
   * Get all drinks
   */
  getDrinks(): Drink[] {
    return [...this.drinks];
  }

  /**
   * Get all foods
   */
  getFoods(): Food[] {
    return [...this.foods];
  }

  /**
   * Get the most recent food before the reference time
   * @returns [foodType, minutesSinceEaten]
   */
  private getMostRecentFood(referenceTime: Date): [FoodType, number] {
    const recentFoods = this.foods.filter(f => f.time <= referenceTime);

    if (recentFoods.length === 0) {
      return ['empty_stomach', Infinity];
    }

    const latestFood = recentFoods[recentFoods.length - 1];
    const minutesElapsed = (referenceTime.getTime() - latestFood.time.getTime()) / (60 * 1000);
    return [latestFood.type, minutesElapsed];
  }

  /**
   * Calculate absorption factor (0-1, where 1 = fully absorbed) for a drink.
   * Based on food state at drink time and time elapsed since drinking.
   */
  private calculateAbsorptionFactor(drinkTime: Date, targetTime: Date = new Date()): number {
    // Time elapsed since drink was consumed (in minutes)
    const minutesSinceDrink = Math.max(0, (targetTime.getTime() - drinkTime.getTime()) / (60 * 1000));

    // Get food state at time of drink
    const [foodType, minutesSinceFood] = this.getMostRecentFood(drinkTime);
    const gastricHalfTime = FOOD_GASTRIC_TIMES[foodType] ?? 90;

    if (gastricHalfTime === 0) {
      // Empty stomach - fast absorption: ~80% in 30 min, ~95% in 60 min
      // Ensure minimum 10% immediate absorption
      const absorption = 0.10 + 0.90 * (1.0 - Math.exp(-minutesSinceDrink / 20));
      return Math.min(1.0, absorption);
    } else {
      // Food delays absorption
      const delayFactor = Math.min(1.0, minutesSinceFood / gastricHalfTime);

      // Adjust absorption rate based on food
      const effectiveAbsorptionTime = minutesSinceDrink * (0.5 + 0.5 * delayFactor);

      // Ensure minimum 10% immediate absorption even with food
      const absorption = 0.10 + 0.90 * (1.0 - Math.exp(-effectiveAbsorptionTime / 30));
      return Math.min(1.0, absorption);
    }
  }

  /**
   * Calculate BAC at a specific time using Widmark equation with food absorption.
   * BAC = [(A × 5.14) / (W × r)] - (0.015 × H)
   */
  calculateBACAtTime(targetTime: Date = new Date()): number {
    if (targetTime < this.startTime) {
      return 0.0;
    }

    // Get Widmark parameters
    const widmarkRatio = WIDMARK_RATIOS[this.profile.sex];
    const bodyWeight = this.profile.weightLbs;

    // Account for metabolism variation
    let eliminationRate = ELIMINATION_RATE;
    if (this.profile.chronicDrinker) {
      eliminationRate *= 1.2; // 20% faster for chronic drinkers
    }

    // Calculate absorption for each drink
    let totalAlcoholAbsorbed = 0.0;

    for (const drink of this.drinks) {
      if (drink.time <= targetTime) {
        // Alcohol in this drink (liquid ounces)
        const alcoholOz = drink.sizeOz * (drink.alcoholPercent / 100);

        // Absorption factor accounts for food effects and time elapsed
        const absorptionFactor = this.calculateAbsorptionFactor(drink.time, targetTime);

        // Food reduces peak BAC (applied separately from absorption timing)
        const [foodType] = this.getMostRecentFood(drink.time);
        const peakReduction = FOOD_ABSORPTION_IMPACT[foodType] ?? 0.0;

        // Effective alcohol = actual alcohol * how much absorbed * food peak reduction
        const effectiveAlcoholOz = alcoholOz * absorptionFactor * (1 - peakReduction * 0.5);

        totalAlcoholAbsorbed += effectiveAlcoholOz;
      }
    }

    // Widmark equation: BAC = [(A × 5.14) / (W × r)] - (0.015 × H)
    let bacFromAbsorption = 0.0;
    if (totalAlcoholAbsorbed > 0) {
      bacFromAbsorption = (totalAlcoholAbsorbed * 5.14) / (bodyWeight * widmarkRatio);
    }

    // Time since first drink (hours)
    const timeSinceFirst = (targetTime.getTime() - this.startTime.getTime()) / (3600 * 1000);
    const elimination = eliminationRate * Math.max(0, timeSinceFirst);

    const bac = Math.max(0.0, bacFromAbsorption - elimination);
    return Math.round(bac * 10000) / 10000; // Round to 4 decimal places
  }

  /**
   * Get current BAC
   */
  getCurrentBAC(): number {
    return this.calculateBACAtTime(new Date());
  }

  /**
   * Generate BAC values for timeline visualization
   * @param hours Number of hours to project
   * @param fromNow If true, start from current time (future projection)
   */
  getBACTimeline(hours: number = 6, fromNow: boolean = true): TimelinePoint[] {
    const timeline: TimelinePoint[] = [];

    const start = fromNow ? new Date() : this.startTime;
    const endTime = new Date(start.getTime() + hours * 60 * 60 * 1000);

    // Sample every 5 minutes
    let currentTime = new Date(start);
    while (currentTime <= endTime) {
      const bac = this.calculateBACAtTime(currentTime);
      timeline.push({ time: new Date(currentTime), bac });
      currentTime = new Date(currentTime.getTime() + 5 * 60 * 1000);
    }

    return timeline;
  }

  /**
   * Find peak BAC and when it occurs
   */
  getPeakBAC(): BACPeak {
    const timeline = this.getBACTimeline(6, false);
    let peakBac = 0.0;
    let peakTime = this.startTime;

    for (const point of timeline) {
      if (point.bac > peakBac) {
        peakBac = point.bac;
        peakTime = point.time;
      }
    }

    return { bac: peakBac, time: peakTime };
  }

  /**
   * Calculate time until BAC drops below threshold
   * @param threshold BAC threshold (default: 0.0 for complete sobriety)
   * @returns Minutes until threshold reached, or null if already below
   */
  getTimeToSobriety(threshold: number = 0.0): number | null {
    const timeline = this.getBACTimeline(24, true);
    const now = new Date();

    for (const point of timeline) {
      if (point.bac <= threshold) {
        return Math.round((point.time.getTime() - now.getTime()) / (60 * 1000));
      }
    }

    return null; // More than 24 hours
  }

  /**
   * Calculate time until BAC reaches legal limit (0.08%)
   * @returns Minutes until legal limit reached, or null if won't reach
   */
  getTimeToLegalLimit(): number | null {
    const timeline = this.getBACTimeline(6, true);
    const now = new Date();

    for (const point of timeline) {
      if (point.bac >= 0.08) {
        return Math.round((point.time.getTime() - now.getTime()) / (60 * 1000));
      }
    }

    return null; // Won't reach legal limit
  }

  /**
   * Get time when BAC will drop below legal limit
   * @returns Date when safe to drive, or null if already legal
   */
  getTimeWhenLegal(): Date | null {
    const currentBAC = this.getCurrentBAC();
    if (currentBAC < 0.08) {
      return null; // Already legal
    }

    const timeline = this.getBACTimeline(24, true);
    for (const point of timeline) {
      if (point.bac < 0.08) {
        return point.time;
      }
    }

    return null;
  }

  /**
   * Get impairment description and legal status for BAC level
   */
  getImpairmentLevel(bac?: number): ImpairmentLevel {
    const currentBac = bac ?? this.getCurrentBAC();

    // Find appropriate level (iterate in reverse to find highest matching threshold)
    for (let i = IMPAIRMENT_LEVELS.length - 1; i >= 0; i--) {
      if (currentBac >= IMPAIRMENT_LEVELS[i].threshold) {
        return IMPAIRMENT_LEVELS[i];
      }
    }

    return IMPAIRMENT_LEVELS[0];
  }

  /**
   * Get full BAC result with impairment level
   */
  getBACResult(): BACResult {
    const bac = this.getCurrentBAC();
    return {
      bac,
      impairmentLevel: this.getImpairmentLevel(bac),
      timestamp: new Date(),
    };
  }

  /**
   * Reset all data for new scenario
   */
  clearScenario(): void {
    this.drinks = [];
    this.foods = [];
    this.startTime = new Date();
  }

  /**
   * Check if there are any drinks logged
   */
  hasDrinks(): boolean {
    return this.drinks.length > 0;
  }

  /**
   * Get start time of drinking session
   */
  getStartTime(): Date {
    return new Date(this.startTime);
  }
}

/**
 * Create a new BAC calculator instance
 */
export function createBACCalculator(profile?: Partial<UserProfile>): BACCalculator {
  return new BACCalculator(profile);
}

/**
 * Format BAC as percentage string
 */
export function formatBAC(bac: number): string {
  return `${(bac * 100).toFixed(2)}%`;
}

/**
 * Format BAC as decimal string (e.g., "0.08")
 */
export function formatBACDecimal(bac: number): string {
  return bac.toFixed(3);
}

/**
 * Format time duration in minutes to human-readable string
 */
export function formatDuration(minutes: number | null): string {
  if (minutes === null) return 'N/A';
  if (minutes < 60) return `${Math.round(minutes)} min`;

  const hours = Math.floor(minutes / 60);
  const mins = Math.round(minutes % 60);

  if (mins === 0) return `${hours}h`;
  return `${hours}h ${mins}m`;
}

/**
 * Calculate standard drinks equivalent
 * One standard drink = 0.6 oz of pure alcohol
 */
export function calculateStandardDrinks(sizeOz: number, alcoholPercent: number): number {
  const pureAlcoholOz = sizeOz * (alcoholPercent / 100);
  return Math.round((pureAlcoholOz / 0.6) * 10) / 10;
}

File: BAC_Simulator.app/Contents/Resources/bac_calculator.py
"""
BAC Calculation Engine - Widmark Equation with Food Absorption Model
Scientifically accurate blood alcohol content simulator
"""
import math
from datetime import datetime, timedelta
from typing import Dict, List, Tuple

class BACCalculator:
    """
    Core BAC calculation engine using modified Widmark equation:
    BAC = [(A × 5.14) / (W × r)] - (0.015 × H)
    """

    # Widmark distribution ratios
    WIDMARK_RATIOS = {
        'male': 0.73,
        'female': 0.66
    }

    # Elimination rate: % BAC per hour (15 mg/100mL per hour = ~0.015%)
    ELIMINATION_RATE = 0.015  # %/hour

    # Food gastric emptying times (minutes) - half-life of stomach content
    FOOD_GASTRIC_TIMES = {
        'empty_stomach': 0,
        'water': 15,
        'light_snack': 60,
        'light_meal': 90,
        'moderate_meal': 120,
        'full_meal': 150,
        'high_fat_meal': 180
    }

    # Food absorption impact (% reduction in peak BAC)
    FOOD_ABSORPTION_IMPACT = {
        'empty_stomach': 0,      # 0% reduction
        'water': 0.05,           # 5% reduction
        'light_snack': 0.20,     # 20% reduction
        'light_meal': 0.30,      # 30% reduction
        'moderate_meal': 0.40,   # 40% reduction
        'full_meal': 0.45,       # 45% reduction
        'high_fat_meal': 0.60    # 60% reduction
    }

    # Standard drink sizes and alcohol content
    STANDARD_DRINKS = {
        'beer_light': {'oz': 12, 'alcohol_percent': 4.2},
        'beer_regular': {'oz': 12, 'alcohol_percent': 5.0},
        'beer_ipa': {'oz': 12, 'alcohol_percent': 6.5},
        'beer_stout': {'oz': 12, 'alcohol_percent': 7.0},
        'wine_light': {'oz': 5, 'alcohol_percent': 11.0},
        'wine_red': {'oz': 5, 'alcohol_percent': 13.5},
        'wine_fortified': {'oz': 3, 'alcohol_percent': 20.0},
        'spirits': {'oz': 1.5, 'alcohol_percent': 40.0},
        'mixed_drink': {'oz': 1.5, 'alcohol_percent': 40.0},
    }

    def __init__(self):
        self.drinks_timeline = []  # List of {time, drink_type, size_oz, alcohol_percent}
        self.food_timeline = []    # List of {time, food_type}
        self.profile = {
            'sex': 'male',
            'weight_lbs': 180,
            'height_inches': 70,
            'age': 30,
            'chronic_drinker': False,
            'medications': []
        }
        self.start_time = datetime.now()

    def set_profile(self, sex: str, weight_lbs: float, age: int = 30,
                   chronic_drinker: bool = False):
        """Set user profile for BAC calculations"""
        self.profile = {
            'sex': sex.lower(),
            'weight_lbs': float(weight_lbs),
            'age': int(age),
            'chronic_drinker': chronic_drinker,
            'medications': []
        }

    def add_food(self, time: datetime, food_type: str):
        """Add food consumed to timeline"""
        food_type = food_type.lower()
        if food_type not in self.FOOD_GASTRIC_TIMES:
            food_type = 'light_meal'  # Default
        self.food_timeline.append({'time': time, 'type': food_type})
        self.food_timeline.sort(key=lambda x: x['time'])

    def add_drink(self, time: datetime, drink_type: str, size_oz: float = None,
                 alcohol_percent: float = None, quantity: int = 1):
        """Add drink(s) consumed to timeline"""
        drink_type_lower = drink_type.lower()

        if drink_type_lower in self.STANDARD_DRINKS:
            std_drink = self.STANDARD_DRINKS[drink_type_lower]
            size_oz = size_oz or std_drink['oz']
            alcohol_percent = alcohol_percent or std_drink['alcohol_percent']
        else:
            size_oz = size_oz or 12
            alcohol_percent = alcohol_percent or 5.0

        for i in range(quantity):
            drink_time = time + timedelta(minutes=i*30)  # Spread drinks 30 min apart
            self.drinks_timeline.append({
                'time': drink_time,
                'type': drink_type,
                'size_oz': float(size_oz),
                'alcohol_percent': float(alcohol_percent)
            })

        self.drinks_timeline.sort(key=lambda x: x['time'])

    def get_most_recent_food(self, reference_time: datetime) -> Tuple[str, float]:
        """
        Get the most recent food before the reference time and minutes elapsed.
        Returns (food_type, minutes_since_eaten)
        """
        recent_foods = [f for f in self.food_timeline if f['time'] <= reference_time]

        if not recent_foods:
            return 'empty_stomach', float('inf')

        latest_food = recent_foods[-1]
        minutes_elapsed = (reference_time - latest_food['time']).total_seconds() / 60
        return latest_food['type'], minutes_elapsed

    def calculate_absorption_factor(self, drink_time: datetime, target_time: datetime = None) -> float:
        """
        Calculate absorption factor (0-1, where 1 = fully absorbed) for a drink.
        Based on food state at drink time and time elapsed since drinking.

        Args:
            drink_time: When the drink was consumed
            target_time: Time at which to calculate absorption (default: now)
        """
        if target_time is None:
            target_time = datetime.now()

        # Time elapsed since drink was consumed (in minutes)
        minutes_since_drink = max(0, (target_time - drink_time).total_seconds() / 60)

        # Get food state at time of drink
        food_type, minutes_since_food = self.get_most_recent_food(drink_time)
        gastric_half_time = self.FOOD_GASTRIC_TIMES.get(food_type, 90)

        if gastric_half_time == 0:  # Empty stomach
            # Fast absorption: ~80% in 30 min, ~95% in 60 min
            # Ensure minimum 10% immediate absorption
            absorption = 0.10 + 0.90 * (1.0 - math.exp(-minutes_since_drink / 20))
            return min(1.0, absorption)
        else:
            # Food delays absorption
            # Calculate effective delay based on how full stomach was
            delay_factor = min(1.0, minutes_since_food / gastric_half_time)

            # Adjust absorption rate based on food
            # Full stomach: slower absorption, takes ~90-120 min to absorb fully
            # As stomach empties, absorption speeds up
            effective_absorption_time = minutes_since_drink * (0.5 + 0.5 * delay_factor)

            # Ensure minimum 10% immediate absorption even with food
            absorption = 0.10 + 0.90 * (1.0 - math.exp(-effective_absorption_time / 30))
            return min(1.0, absorption)

    def calculate_bac_at_time(self, target_time: datetime = None) -> float:
        """
        Calculate BAC at a specific time using Widmark equation with food absorption.
        BAC = [(A × 5.14) / (W × r)] - (0.015 × H)
        """
        if target_time is None:
            target_time = datetime.now()

        if target_time < self.start_time:
            return 0.0

        # Get Widmark parameters
        widmark_ratio = self.WIDMARK_RATIOS.get(self.profile['sex'], 0.73)
        body_weight = self.profile['weight_lbs']

        # Account for metabolism variation
        elimination_rate = self.ELIMINATION_RATE
        if self.profile['chronic_drinker']:
            elimination_rate *= 1.2  # 20% faster for chronic drinkers

        # Calculate absorption for each drink
        total_alcohol_absorbed = 0.0

        for drink in self.drinks_timeline:
            if drink['time'] <= target_time:
                # Alcohol in this drink (liquid ounces)
                alcohol_oz = drink['size_oz'] * (drink['alcohol_percent'] / 100)

                # Absorption factor accounts for food effects and time elapsed
                absorption_factor = self.calculate_absorption_factor(drink['time'], target_time)

                # Food reduces peak BAC (applied separately from absorption timing)
                food_type, _ = self.get_most_recent_food(drink['time'])
                peak_reduction = self.FOOD_ABSORPTION_IMPACT.get(food_type, 0.0)

                # Effective alcohol = actual alcohol * how much absorbed * food peak reduction
                # Note: absorption_factor handles WHEN alcohol is absorbed
                # peak_reduction handles HOW MUCH of the peak is reduced
                effective_alcohol_oz = alcohol_oz * absorption_factor * (1 - peak_reduction * 0.5)

                # Add to total
                total_alcohol_absorbed += effective_alcohol_oz

        # Widmark equation: BAC = [(A × 5.14) / (W × r)] - (0.015 × H)
        if total_alcohol_absorbed > 0:
            bac_from_absorption = (total_alcohol_absorbed * 5.14) / (body_weight * widmark_ratio)
        else:
            bac_from_absorption = 0.0

        # Time since first drink
        time_since_first = (target_time - self.start_time).total_seconds() / 3600
        elimination = elimination_rate * max(0, time_since_first)

        bac = max(0.0, bac_from_absorption - elimination)
        return round(bac, 4)

    def get_bac_timeline(self, hours: int = 6, from_now: bool = True) -> List[Tuple[datetime, float]]:
        """
        Generate BAC values for timeline visualization.

        Args:
            hours: Number of hours to project
            from_now: If True, start from current time (future projection).
                      If False, start from drinking start time (full history).
        """
        timeline = []

        if from_now:
            # Future projection from current time
            start = datetime.now()
        else:
            # Full history from when drinking started
            start = self.start_time

        end_time = start + timedelta(hours=hours)

        # Sample every 5 minutes
        current_time = start
        while current_time <= end_time:
            bac = self.calculate_bac_at_time(current_time)
            timeline.append((current_time, bac))
            current_time += timedelta(minutes=5)

        return timeline

    def get_peak_bac(self) -> Tuple[float, datetime]:
        """Find peak BAC and when it occurs"""
        timeline = self.get_bac_timeline(hours=6)
        peak_bac = 0.0
        peak_time = self.start_time

        for time, bac in timeline:
            if bac > peak_bac:
                peak_bac = bac
                peak_time = time

        return peak_bac, peak_time

    def get_time_to_sobriety(self, threshold: float = 0.0) -> timedelta:
        """Calculate time until BAC drops below threshold"""
        timeline = self.get_bac_timeline(hours=24)

        for time, bac in timeline:
            if bac <= threshold:
                return time - self.start_time

        return timedelta(hours=24)

    def get_time_to_legal_limit(self) -> timedelta:
        """Calculate time until BAC reaches 0.08% (legal limit)"""
        timeline = self.get_bac_timeline(hours=6)

        for time, bac in timeline:
            if bac >= 0.08:
                return time - self.start_time

        return None

    def get_impairment_level(self, bac: float = None) -> Dict:
        """Get impairment description and legal status for BAC level"""
        if bac is None:
            bac = self.calculate_bac_at_time()

        impairment_levels = [
            {
                'threshold': 0.0,
                'level': 'Sober',
                'description': 'No detectable impairment',
                'color': 'green',
                'fitness_to_drive': 'YES',
                'legal_status': 'LEGAL'
            },
            {
                'threshold': 0.02,
                'level': 'Minimal Impairment',
                'description': 'Slight warmth, mild euphoria, minimal coordination loss',
                'color': 'lightgreen',
                'fitness_to_drive': 'YES',
                'legal_status': 'LEGAL'
            },
            {
                'threshold': 0.05,
                'level': 'Mild Impairment',
                'description': 'Reduced concentration, slower reaction time, slight loss of coordination',
                'color': 'yellow',
                'fitness_to_drive': 'CAUTION',
                'legal_status': 'LEGAL'
            },
            {
                'threshold': 0.08,
                'level': 'Moderate Impairment',
                'description': 'Legal limit reached - DUI threshold for standard drivers',
                'color': 'orange',
                'fitness_to_drive': 'NO',
                'legal_status': 'ILLEGAL - DUI'
            },
            {
                'threshold': 0.15,
                'level': 'Severe Impairment',
                'description': 'Enhanced DUI threshold in Tennessee (7+ day jail for first offense)',
                'color': 'darkorange',
                'fitness_to_drive': 'NO',
                'legal_status': 'ILLEGAL - ENHANCED DUI'
            },
            {
                'threshold': 0.20,
                'level': 'Very Severe Impairment',
                'description': 'Major loss of motor control, risk of blackouts, danger of poisoning',
                'color': 'red',
                'fitness_to_drive': 'NO',
                'legal_status': 'DANGEROUS - MEDICAL RISK'
            },
            {
                'threshold': 0.30,
                'level': 'Extreme Intoxication',
                'description': 'Severe loss of consciousness, risk of death, medical emergency',
                'color': 'darkred',
                'fitness_to_drive': 'NO',
                'legal_status': 'LIFE-THREATENING'
            }
        ]

        # Find appropriate level
        for level in reversed(impairment_levels):
            if bac >= level['threshold']:
                return level

        return impairment_levels[0]

    def clear_scenario(self):
        """Reset all data for new scenario"""
        self.drinks_timeline = []
        self.food_timeline = []
        self.start_time = datetime.now()
